File: src/types/SexPractice.ts
import * as DB from "tsondb/schema/dsl"
import { NestedTranslationMap } from "./Locale.js"
import { src } from "./source/_PublicationRef.js"

export const SexPractice = DB.Entity(import.meta.url, {
  name: "SexPractice",
  namePlural: "SexPractices",
  type: () =>
    DB.Object({
      src,
      translations: NestedTranslationMap(
        DB.Required,
        "SexPractice",
        DB.Object({
          name: DB.Required({
            comment: "The sex practice’s name.",
            type: DB.String({ minLength: 1 }),
          }),
          rules: DB.Required({
            comment: "The rules of the sex practice.",
            type: DB.String({ minLength: 1 }),
          }),
          duration: DB.Required({
            comment: "How long a round of this sex practice takes.",
            type: DB.String({ minLength: 1 }),
          }),
          prerequisites: DB.Optional({
            comment:
              "Prerequisites of participants and environment. Do not specify if the sex practice has no prerequisites.",
            type: DB.String({ minLength: 1 }),
          }),
          failed: DB.Required({
            comment: "Effects of a failed *Seduction* check.",
            type: DB.String({ minLength: 1 }),
          }),
        }),
      ),
    }),
  instanceDisplayName: {},
  uniqueConstraints: [
    {
      entityMapKeyPath: "translations",
      keyPathInEntityMap: "name",
    },
  ],
})

File: Sources/OptolithDatabaseSchema/Types/SexPractice.swift
import FileDB

@Model
public struct SexPractice {
    /// The publications where you can find the entry.
    @MinItems(1)
    let src: [PublicationRef]

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [Locale.ID: Translation]

    @Embedded
    struct Translation {  // StateTranslation
        /// The sex practice’s name.
        @MinLength(1)
        let name: String

        /// The rules of the sex practice.
        @MinLength(1)
        let rules: String

        /// How long a round of this sex practice takes.
        @MinLength(1)
        let duration: String

        /// Prerequisites of participants and environment. Do not specify if the sex practice has no prerequisites.
        @MinLength(1)
        let prerequisites: String?

        /// Effects of a failed *Seduction* check.
        @MinLength(1)
        let failed: String
    }
}
